lib/elixir/src/elixir_bitstring.erl

```diff
@@ -244,12 +244,19 @@ validate_spec(signed, [])    -> {sign, signed};
 validate_spec(unsigned, [])  -> {sign, unsigned};
 validate_spec(_, _)          -> none.
 
-expand_spec_arg(Expr, S, _OriginalS, E) when is_atom(Expr); is_integer(Expr) ->
+expand_spec_arg(_, Expr, S, _OriginalS, E) when is_atom(Expr); is_integer(Expr) ->
   {Expr, S, E};
-expand_spec_arg(Expr, S, _OriginalS, #{context := match} = E) ->
-  {EExpr, SE, EE} = elixir_expand:expand(Expr, S#elixir_ex{prematch=raise}, E#{context := nil}),
+expand_spec_arg(size, Expr, S, _OriginalS, #{context := match} = E) ->
+  {EExpr, SE, EE} = elixir_expand:expand(Expr, S#elixir_ex{prematch=raise, bitsize=true}, E#{context := guard}),
+  {EExpr, SE#elixir_ex{prematch=S#elixir_ex.prematch, bitsize=false}, EE#{context := match}};
+expand_spec_arg(size, Expr, S, OriginalS, E) ->
+  NewS = elixir_env:reset_read(S, OriginalS),
+  {EExpr, SE, EE} = elixir_expand:expand(Expr, NewS#elixir_ex{bitsize=true}, E#{context := guard}),
+  {EExpr, SE#elixir_ex{bitsize=false}, EE#{context := nil}};
```

Comment:
We should remove all of this:

```suggestion
```

And add a test that `<<255::size(hd(List.flatten([3])))>>` works :)

Reply:
I applied this suggestion but several tests in `lib/elixir/test/elixir/kernel/expansion_test.exs` and `lib/elixir/test/elixir/kernel/binary_test.exs` are now broken (as I expected).

So I just reverted the addition of the Kind argument (because I added it, it was there before), and I found all tests still passes. I think this is your intention, but I leave you to check.

Reply:
> And add a test that `<<255::size(hd(List.flatten([3])))>>` works :)

Note this fails because it is a remote call to the `List` module, not allowed in guards. I added a similar test only using `hd`.